File: electron/control/finoraInstallationBindingService.ts
/* ===========================================================
   FINORA ENTERPRISE OS™

   WINDOWS INSTALLATION BINDING SERVICE

   MODULE  : Native Control
   LAYER   : Electron Main
   VERSION : 1.1
   STATUS  : Production Foundation

   RESPONSIBILITY:

   - Reconcile existing Control Store installation identity
     with native installation-binding identity
   - Preserve legacy installationId when creating the first
     binding vault
   - Create a fresh native installationId only when neither
     identity exists
   - Fail closed if Control Store and binding vault disagree
   - Expose only public binding metadata to Electron-main code
   - Sign installation-enrollment proof internally

   SECURITY:

   - No IPC.
   - No preload.
   - No renderer.
   - No raw private-key getter.
   - No Control Center signing authority.
   - No Business Date.
=========================================================== */

import {
  generateFinoraWindowsInstallationBindingMaterial,
  signFinoraInstallationEnrollmentCanonicalValue,
  toFinoraWindowsInstallationBindingPublic,
} from "./finoraInstallationBindingCrypto.js";

import {
  reconcileFinoraInstallationBinding,
} from "./finoraInstallationBindingReconciliation.js";

import {
  loadFinoraInstallationBindingVault,
  persistNewFinoraInstallationBindingVault,
} from "./finoraInstallationBindingVault.js";

import {
  getFinoraInstallationIdentity,
} from "./finoraControlStore.js";

import type {
  FinoraWindowsInstallationBindingMaterial,
  FinoraWindowsInstallationBindingPublic,
} from "./finoraInstallationBindingCrypto.js";

// ============================================================
// SERIALIZED CREATION
// ============================================================

let installationBindingQueue:
  Promise<void> =
    Promise.resolve();

// ============================================================
// CONTROL STORE INSTALLATION ID
// ============================================================

async function loadControlInstallationId():
  Promise<
    string |
    undefined
  > {

  const result =
    await getFinoraInstallationIdentity();

  if (!result.success) {

    throw new Error(
      result.error ??
        "Unable to load FINORA Control Store installation identity.",
    );
  }

  return result.data
    ?.installationId;
}

// ============================================================
// CONSISTENCY GUARD
// ============================================================

async function assertBindingConsistency(
  material:
    FinoraWindowsInstallationBindingMaterial,
): Promise<void> {

  const controlInstallationId =
    await loadControlInstallationId();

  const reconciliation =
    reconcileFinoraInstallationBinding(
      controlInstallationId,
      material.installationId,
    );

  if (!reconciliation.accepted) {

    throw new Error(
      reconciliation.error,
    );
  }

  if (
    reconciliation.action !==
      "USE_EXISTING_BINDING"
  ) {

    throw new Error(
      "FINORA installation binding consistency state is invalid.",
    );
  }
}

// ============================================================
// INTERNAL ENSURE
// ============================================================

async function ensureBindingInternal():
  Promise<
    FinoraWindowsInstallationBindingPublic
  > {

  const controlInstallationId =
    await loadControlInstallationId();

  const existing =
    await loadFinoraInstallationBindingVault();

  const reconciliation =
    reconcileFinoraInstallationBinding(
      controlInstallationId,
      existing?.installationId,
    );

  if (!reconciliation.accepted) {

    throw new Error(
      reconciliation.error,
    );
  }

  if (
    reconciliation.action ===
      "USE_EXISTING_BINDING"
  ) {

    if (!existing) {

      throw new Error(
        "FINORA installation binding reconciliation expected an existing native binding.",
      );
    }

    return toFinoraWindowsInstallationBindingPublic(
      existing,
    );
  }

  const generated =
    generateFinoraWindowsInstallationBindingMaterial(
      new Date(),
      reconciliation.installationId,
    );

  try {

    await persistNewFinoraInstallationBindingVault(
      generated,
    );

    return toFinoraWindowsInstallationBindingPublic(
      generated,
    );

  } catch (
    error
  ) {

    /*
     * Another native path may have established the immutable
     * vault first. Only accept that authoritative vault if it
     * reconciles with the Control Store identity.
     */
    const authoritative =
      await loadFinoraInstallationBindingVault();

    if (!authoritative) {
      throw error;
    }

    const authoritativeReconciliation =
      reconcileFinoraInstallationBinding(
        controlInstallationId,
        authoritative.installationId,
      );

    if (
      !authoritativeReconciliation.accepted ||
      authoritativeReconciliation.action !==
        "USE_EXISTING_BINDING"
    ) {

      throw new Error(
        !authoritativeReconciliation.accepted
          ? authoritativeReconciliation.error
          : "FINORA authoritative installation binding state is invalid.",
      );
    }

    return toFinoraWindowsInstallationBindingPublic(
      authoritative,
    );
  }
}

// ============================================================
// ENSURE
// ============================================================

export function ensureFinoraWindowsInstallationBinding():
  Promise<
    FinoraWindowsInstallationBindingPublic
  > {

  const operation =
    installationBindingQueue.then(
      () =>
        ensureBindingInternal(),
    );

  installationBindingQueue =
    operation.then(
      () =>
        undefined,
      () =>
        undefined,
    );

  return operation;
}

// ============================================================
// READ PUBLIC BINDING
// ============================================================

export async function getFinoraWindowsInstallationBinding():
  Promise<
    FinoraWindowsInstallationBindingPublic |
    undefined
  > {

  const material =
    await loadFinoraInstallationBindingVault();

  if (!material) {
    return undefined;
  }

  await assertBindingConsistency(
    material,
  );

  return toFinoraWindowsInstallationBindingPublic(
    material,
  );
}

// ============================================================
// INTERNAL ENROLLMENT POSSESSION PROOF
// ============================================================

export async function signFinoraWindowsInstallationEnrollment(
  canonicalEnrollmentPayload:
    string,
): Promise<string> {

  const material =
    await loadFinoraInstallationBindingVault();

  if (!material) {

    throw new Error(
      "FINORA Windows installation binding has not been initialized.",
    );
  }

  await assertBindingConsistency(
    material,
  );

  return signFinoraInstallationEnrollmentCanonicalValue(
    canonicalEnrollmentPayload,
    material,
  );
}

// ============================================================
// END
// ============================================================

File: android/app/src/main/java/com/finora/enterprise/control/FinoraInstallationBindingService.java
package com.finora.enterprise.control;

/* ===========================================================
   FINORA ENTERPRISE OS™

   ANDROID INSTALLATION BINDING SERVICE

   MODULE  : Native Control
   LAYER   : Android Native
   VERSION : 1.0
   STATUS  : Production Foundation

   RESPONSIBILITY:

   - Read existing Control Store installationId
   - Reconcile it with Android Keystore binding identity
   - Preserve legacy installation identity
   - Generate fresh installation identity only when required
   - Fail closed on mismatch
   - Provide internal enrollment signing capability

   SECURITY:

   - Native only.
   - No PluginMethod.
   - No WebView.
   - No private-key getter.
   - No Control Center signing authority.
   - No Business Date.
=========================================================== */

import android.content.Context;

import java.util.UUID;

import org.json.JSONObject;

public final class FinoraInstallationBindingService {

    private static final Object LOCK =
        new Object();

    private final FinoraControlStore controlStore;

    public FinoraInstallationBindingService(
        Context context
    ) {

        if (context == null) {
            throw new IllegalArgumentException(
                "Android context is required."
            );
        }

        controlStore =
            new FinoraControlStore(
                context.getApplicationContext()
            );
    }

    // ========================================================
    // ENSURE
    // ========================================================

    public FinoraInstallationBindingCrypto.PublicBinding ensure()
        throws Exception {

        synchronized (LOCK) {

            String controlInstallationId =
                loadControlInstallationId();

            FinoraInstallationBindingCrypto.PublicBinding existing =
                FinoraInstallationBindingCrypto
                    .loadExisting();

            FinoraInstallationBindingReconciliation.Decision decision =
                FinoraInstallationBindingReconciliation
                    .reconcile(
                        controlInstallationId,
                        existing == null
                            ? null
                            : existing.installationId
                    );

            if (!decision.accepted) {
                throw new IllegalStateException(
                    decision.error
                );
            }

            if (
                FinoraInstallationBindingReconciliation
                    .ACTION_USE_EXISTING
                    .equals(
                        decision.action
                    )
            ) {

                if (existing == null) {
                    throw new IllegalStateException(
                        "FINORA Android installation reconciliation expected an existing native binding."
                    );
                }

                return existing;
            }

            if (
                !FinoraInstallationBindingReconciliation
                    .ACTION_CREATE
                    .equals(
                        decision.action
                    )
            ) {
                throw new IllegalStateException(
                    "FINORA Android installation reconciliation action is invalid."
                );
            }

            String installationId =
                decision.installationId;

            if (
                installationId == null
            ) {
                installationId =
                    "FINORA-INSTALLATION-" +
                    UUID.randomUUID();
            }

            FinoraInstallationBindingCrypto.PublicBinding generated =
                FinoraInstallationBindingCrypto
                    .generate(
                        installationId
                    );

            /*
             * Re-read Control Store after native key generation.
             * If another native startup path established an
             * incompatible installation identity, fail closed.
             */
            String authoritativeControlInstallationId =
                loadControlInstallationId();

            FinoraInstallationBindingReconciliation.Decision finalDecision =
                FinoraInstallationBindingReconciliation
                    .reconcile(
                        authoritativeControlInstallationId,
                        generated.installationId
                    );

            if (
                !finalDecision.accepted ||
                !FinoraInstallationBindingReconciliation
                    .ACTION_USE_EXISTING
                    .equals(
                        finalDecision.action
                    )
            ) {
                throw new IllegalStateException(
                    finalDecision.error != null
                        ? finalDecision.error
                        : "FINORA Android installation binding final consistency check failed."
                );
            }

            return generated;
        }
    }

    // ========================================================
    // READ PUBLIC BINDING
    // ========================================================

    public FinoraInstallationBindingCrypto.PublicBinding get()
        throws Exception {

        synchronized (LOCK) {

            FinoraInstallationBindingCrypto.PublicBinding binding =
                FinoraInstallationBindingCrypto
                    .loadExisting();

            if (binding == null) {
                return null;
            }

            assertConsistency(
                binding
            );

            return binding;
        }
    }

    // ========================================================
    // INTERNAL ENROLLMENT SIGNING
    // ========================================================

    public String signEnrollment(
        String canonicalPayload
    ) throws Exception {

        synchronized (LOCK) {

            FinoraInstallationBindingCrypto.PublicBinding binding =
                FinoraInstallationBindingCrypto
                    .loadExisting();

            if (binding == null) {
                throw new IllegalStateException(
                    "FINORA Android installation binding has not been initialized."
                );
            }

            assertConsistency(
                binding
            );

            return FinoraInstallationBindingCrypto
                .signCanonicalEnrollment(
                    canonicalPayload
                );
        }
    }

    // ========================================================
    // CONSISTENCY
    // ========================================================

    private void assertConsistency(
        FinoraInstallationBindingCrypto.PublicBinding binding
    ) throws Exception {

        String controlInstallationId =
            loadControlInstallationId();

        FinoraInstallationBindingReconciliation.Decision decision =
            FinoraInstallationBindingReconciliation
                .reconcile(
                    controlInstallationId,
                    binding.installationId
                );

        if (
            !decision.accepted ||
            !FinoraInstallationBindingReconciliation
                .ACTION_USE_EXISTING
                .equals(
                    decision.action
                )
        ) {
            throw new IllegalStateException(
                decision.error != null
                    ? decision.error
                    : "FINORA Android installation binding consistency check failed."
            );
        }
    }

    // ========================================================
    // CONTROL STORE INSTALLATION ID
    // ========================================================

    private String loadControlInstallationId()
        throws Exception {

        String raw =
            controlStore.read();

        if (raw == null) {
            return null;
        }

        JSONObject controlPackage =
            new JSONObject(
                raw
            );

        JSONObject installation =
            controlPackage.optJSONObject(
                "installation"
            );

        if (installation == null) {
            return null;
        }

        String installationId =
            installation.optString(
                "installationId",
                ""
            ).trim();

        if (installationId.isEmpty()) {
            throw new IllegalStateException(
                "FINORA Control Store installationId is invalid."
            );
        }

        return installationId;
    }
}